File: ios/Shared/Sources/Backend.swift
import Foundation

/// What answers the app's questions about shopping.
///
/// One protocol with one conformer today -- `API`, over HTTP -- and that is the whole
/// of this change. It is here to make a boundary visible before anything is moved
/// across it.
///
/// ## Why the surface is split in three
///
/// A device kept to itself and a device with a server are meant to be the same app,
/// and today they are the same app in the worst way: standalone is implemented as *a
/// server that fails every request* (`API.reachable`), so every screen goes down an
/// error path and is then told the error is not real. `onDeviceOnly` appears fifty-odd
/// times across eighteen files, and every one of those is a place that leaked out of.
///
/// The fix is not one protocol with everything in it. It is the observation that the
/// two modes differ in **what they offer**, not in **how shopping works**:
///
/// * ``Backend`` is shopping. A list, what is on it, what things are called and how
///   they are grouped. A device on its own can answer every one of these questions
///   from its own database -- which is what makes a local conformer possible, and what
///   would delete those fifty-odd branches.
/// * ``Accounts`` is who may sign in to a server. There is no answer to that without
///   one, and inventing one would be a lie.
/// * ``Sharing`` is who else is on a list. Likewise: a share link names a server, so
///   with none there is no link to make.
///
/// The second and third are **not** things a local conformer should implement badly.
/// They are things that should be *absent*, which is what the screens already do by
/// hiding them -- correctly, because offering to share when there is nobody to share
/// with is a worse app rather than a more uniform one.
///
/// See `docs/review.md` for the sequence this belongs to.
protocol Backend: Sendable {

    // MARK: - Reading

    func lists() async throws -> Listing<List>
    func items(on list: List) async throws -> Listing<Item>
    func units() async throws -> [Unit]
    func tags(orderedFor list: List) async throws -> [Tag]
    func tags(on item: Item, in list: List) async throws -> [Tag]
    func suggestions(matching typed: String, on list: List) async throws -> [String]
    func history(on list: List) async throws -> [RememberedEntry]

    // MARK: - Lists

    func createList(named name: String) async throws -> List
    func rename(_ list: List, to name: String) async throws
    func delete(_ list: List) async throws

    // MARK: - What is on one

    func add(_ line: String, to list: List) async throws
    func setDone(_ item: Item, on list: List, done: Bool) async throws
    /// By id, for a caller holding a queued operation rather than a row -- see the
    /// watch, which ticks things off it has only ever seen as numbers.
    func setDone(itemID: Int64, listID: Int64, done: Bool) async throws
    func update(
        _ item: Item,
        on list: List,
        name: String,
        amount: Double,
        unitID: Int64?
    ) async throws
    func attach(_ tag: Tag, to item: Item, on list: List) async throws
    func detach(_ tag: Tag, from item: Item, on list: List) async throws
    func clearDone(on list: List) async throws
    func delete(_ item: Item, on list: List) async throws

    // MARK: - The categories, which belong to no one list

    func setTagOrder(_ tags: [Tag], on list: List) async throws
    func createTag(named name: String, emoji: String?) async throws -> Tag
    func updateTag(_ tag: Tag, named name: String, emoji: String?) async throws -> Tag
    func deleteTag(_ tag: Tag) async throws

    // MARK: - Somebody else changed something

    /// Remote changes only. Changes made on this device reach the screen through the
    /// database -- see `Cache.observe(list:)` -- so a conformer with no remote has
    /// nothing to report here and should say so by never yielding, rather than by
    /// failing.
    func listChanges() async throws -> AsyncThrowingStream<Void, Error>
    func changes(on list: List) async throws -> AsyncThrowingStream<Void, Error>
}

/// Who may sign in to this server, and who this is.
///
/// Server-only, and deliberately not part of ``Backend``: a device with no server has
/// no account to describe and nobody to admit. A screen that needs this is a screen
/// that should be absent without one.
protocol Accounts: Sendable {
    func whoAmI() async throws -> Me
    func serverAbout() async throws -> ServerAbout
    func admissions() async throws -> [Admitted]
    func admit(_ email: String, note: String?) async throws
    func withdraw(_ email: String) async throws
    func setOwner(_ email: String, _ owner: Bool) async throws
    func setAdmitsAnyone(_ open: Bool) async throws
}

/// Who else is on a list.
///
/// Server-only for the same reason: a share link names a server, so with none there is
/// no link to make and nobody on the other end of one.
protocol Sharing: Sendable {
    func people(on list: List) async throws -> [Person]
    func invite(to list: List, as role: Role) async throws -> String
    func revokeInvites(to list: List) async throws
    func join(withToken token: String) async throws -> List
    func remove(_ person: Person, from list: List) async throws
}

extension Sharing {
    /// The role a share link gives unless somebody says otherwise.
    ///
    /// A protocol requirement cannot carry a default argument, so it is here instead --
    /// which keeps every existing `invite(to:)` call reading as it did.
    func invite(to list: List) async throws -> String {
        try await invite(to: list, as: .editor)
    }
}

// The one conformer, for now. `API` already has every one of these; nothing about it
// changed to satisfy this.
extension API: Backend {}
extension API: Accounts {}
extension API: Sharing {}

File: android/app/src/main/kotlin/com/cernauskas/shoppinglist/data/Backend.kt
package com.cernauskas.shoppinglist.data

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow

/**
 * What answers this app's questions about shopping.
 *
 * The Kotlin half of `ios/Shared/Sources/Backend.swift`, and drawn for the same reason.
 * A device kept to itself and a device with a server are meant to be the same app, and
 * without this they are the same app in the worst way: standalone becomes *a server
 * that fails every request*, so every screen goes down an error path and is then told
 * the error is not real.
 *
 * ## Why the surface is split
 *
 * The two modes differ in **what they offer**, not in **how shopping works**:
 *
 *  * [Backend] is shopping. A list, what is on it, what things are called and how they
 *    are grouped. A device on its own can answer every one of these from its own
 *    database, which is what makes [LocalBackend] possible.
 *  * [Accounts] is who may sign in to a server. There is no answer to that without one.
 *  * [Sharing] is who else is on a list. A share link names a server, so with none
 *    there is no link to make.
 *
 * The second and third are **not** things a local conformer should implement badly.
 * They are things that should be *absent*, which is what the screens do by hiding them —
 * correctly, because offering to share when there is nobody to share with is a worse
 * app rather than a more uniform one. See [Capabilities].
 */
interface Backend {

    // Reading.

    suspend fun lists(): Listing<ShoppingList>
    suspend fun items(list: ShoppingList): Listing<Item>
    suspend fun units(): List<Unit>
    suspend fun tagsOrderedFor(list: ShoppingList): List<Tag>
    suspend fun tagsOn(item: Item, list: ShoppingList): List<Tag>
    suspend fun suggestions(typed: String, list: ShoppingList): List<String>
    suspend fun history(list: ShoppingList): List<RememberedEntry>

    // Lists.

    suspend fun createList(name: String): ShoppingList
    suspend fun rename(list: ShoppingList, name: String)
    suspend fun delete(list: ShoppingList)

    // What is on one.

    suspend fun add(line: String, list: ShoppingList)

    /**
     * `at` is when it *happened*, which is not always now.
     *
     * A tick made away from signal reaches the server whenever the device next has one,
     * and the ordering rules run on the moment somebody decided rather than the moment
     * the news arrived — docs/offline.md. Milliseconds since the epoch, or null for now.
     */
    suspend fun setDone(item: Item, list: ShoppingList, done: Boolean, at: Long? = null)

    suspend fun update(item: Item, list: ShoppingList, name: String, amount: Double, unitId: Long?)
    suspend fun attach(tag: Tag, item: Item, list: ShoppingList)
    suspend fun detach(tag: Tag, item: Item, list: ShoppingList)
    suspend fun clearDone(list: ShoppingList)
    suspend fun delete(item: Item, list: ShoppingList)

    // The categories, which belong to no one list.

    suspend fun setTagOrder(tags: List<Tag>, list: ShoppingList)

    // Somebody else changed something.

    /**
     * The set of lists this person can see is not what they last read.
     *
     * A nudge and never the rows: a watcher told "something moved" and re-reading cannot
     * drift, while one sent the new rows becomes a second opinion about them.
     */
    fun listChanges(): Flow<kotlin.Unit>

    /**
     * This list is not what it was, and *what* about it changed.
     *
     * The kind matters because the answers cost different amounts. A tick means re-read
     * the rows; a category renamed in settings means re-read the vocabulary, which is
     * thirty-one units and twenty-one categories. Told only that "something happened", a
     * screen has to do both — three requests per tick against a server.
     */
    fun changes(list: ShoppingList): Flow<Nudge>

    // What a screen can ask about the backend itself.

    /**
     * Whether the last attempt to reach the far end got there.
     *
     * The difference between "you have no lists" and "I could not find out", which is
     * the bug the cache was built for. A screen reads this rather than inferring it from
     * an error, because a backend that answers from its own database raises none.
     */
    val reachable: Boolean get() = true

    /** How much this backend is holding that has not reached where it is going. */
    val pending: Int get() = 0

    /** The rows on this list with work that has not been sent. */
    suspend fun unsent(list: ShoppingList): Set<String> = emptySet()

    /**
     * Sends whatever is waiting, and says what became of it.
     *
     * Only the losses are worth showing: "three changes sent" is news about plumbing,
     * while "the thing you crossed off had been deleted" is news about the list.
     */
    suspend fun sync(): SyncReport = SyncReport()
}

/** What moved. */
enum class Nudge {
    /** What is on this list: something added, ticked off, corrected, filed or removed. */
    ROWS,

    /**
     * The categories themselves — renamed, added, removed, or reordered for this list.
     * Global, and changed from a screen that belongs to no list.
     */
    CATEGORIES,
}

/** What became of a queue. */
data class SyncReport(
    val sent: Int = 0,
    val waiting: Int = 0,
    /** Something was refused and will not retry itself. The one state worth interrupting for. */
    val refused: Boolean = false,
    /** Changes that can never land — a tick against a row somebody else deleted. */
    val lost: List<String> = emptyList(),
)

/**
 * Who may sign in to this server, and who this is.
 *
 * Server-only, and deliberately not part of [Backend]: a device with no server has no
 * account to describe and nobody to admit. A screen that needs this is a screen that
 * should be absent without one.
 */
interface Accounts {
    suspend fun whoAmI(): Me
    suspend fun serverAbout(): ServerAbout
    suspend fun admissions(): List<Admitted>
    suspend fun admit(email: String, note: String?)
    suspend fun withdraw(email: String)
    suspend fun setOwner(email: String, owner: Boolean)
    suspend fun setAdmitsAnyone(open: Boolean)
}

/**
 * Who else is on a list.
 *
 * Server-only for the same reason: a share link names a server, so with none there is no
 * link to make and nobody on the other end of one.
 */
interface Sharing {
    suspend fun people(list: ShoppingList): List<Person>
    suspend fun invite(list: ShoppingList): String
    suspend fun revokeInvites(list: ShoppingList)
    suspend fun join(token: String): ShoppingList
    suspend fun remove(person: Person, list: ShoppingList)
}

/**
 * A backend that answers nothing, for a screen built before there is one.
 *
 * Not a mode and not a fallback — composition roots hold a backend as state, and this is
 * what "not built yet" looks like for the one frame before it is. Every read is empty
 * and every write is ignored, which is the honest answer to a question asked too early.
 */
object NoBackend : Backend {
    override suspend fun lists() = Listing<ShoppingList>(emptyList(), 0, false)
    override suspend fun items(list: ShoppingList) = Listing<Item>(emptyList(), 0, false)
    override suspend fun units() = emptyList<Unit>()
    override suspend fun tagsOrderedFor(list: ShoppingList) = emptyList<Tag>()
    override suspend fun tagsOn(item: Item, list: ShoppingList) = emptyList<Tag>()
    override suspend fun suggestions(typed: String, list: ShoppingList) = emptyList<String>()
    override suspend fun history(list: ShoppingList) = emptyList<RememberedEntry>()
    override suspend fun createList(name: String) = error("no backend")
    override suspend fun rename(list: ShoppingList, name: String) {}
    override suspend fun delete(list: ShoppingList) {}
    override suspend fun add(line: String, list: ShoppingList) {}
    override suspend fun setDone(item: Item, list: ShoppingList, done: Boolean, at: Long?) {}
    override suspend fun update(
        item: Item,
        list: ShoppingList,
        name: String,
        amount: Double,
        unitId: Long?,
    ) {}
    override suspend fun attach(tag: Tag, item: Item, list: ShoppingList) {}
    override suspend fun detach(tag: Tag, item: Item, list: ShoppingList) {}
    override suspend fun clearDone(list: ShoppingList) {}
    override suspend fun delete(item: Item, list: ShoppingList) {}
    override suspend fun setTagOrder(tags: List<Tag>, list: ShoppingList) {}
    override fun listChanges(): Flow<kotlin.Unit> = emptyFlow()
    override fun changes(list: ShoppingList): Flow<Nudge> = emptyFlow()
}
